File: CS/UnboundListView.Module/ShowDuplicateBooksController.cs
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;
using DevExpress.Xpo;

namespace UnboundListView.Module {
    public partial class ShowDuplicateBooksController : ViewController {
        public ShowDuplicateBooksController() {
            PopupWindowShowAction showDuplicatesAction = new PopupWindowShowAction(this, "Show Duplicate Books", "View");
            showDuplicatesAction.CustomizePopupWindowParams += new CustomizePopupWindowParamsEventHandler(showDuplicatesAction_CustomizePopupWindowParams);
            TargetViewType = ViewType.ListView;
            TargetObjectType = typeof(Book);
        }

        void showDuplicatesAction_CustomizePopupWindowParams(object sender, CustomizePopupWindowParamsEventArgs e) {
            Dictionary<string, int> dictionary = new Dictionary<string, int>();
            foreach(Book book in ((ListView)View).CollectionSource.List) {
                if(!string.IsNullOrEmpty(book.Title)) {
                    if(dictionary.ContainsKey(book.Title)) {
                        dictionary[book.Title]++;
                    }
                    else dictionary.Add(book.Title, 1);
                }
            }
            ObjectSpace objectSpace = Application.CreateObjectSpace();
            CollectionSource collectionSource = new CollectionSource(objectSpace, typeof(Duplicate));
            if((collectionSource.Collection as XPBaseCollection) != null) {
                ((XPBaseCollection)collectionSource.Collection).LoadingEnabled = false;
            }
            foreach(KeyValuePair<string, int> duplicateRecord in dictionary) {
                if(duplicateRecord.Value > 1) {
                    Duplicate duplicate = new Duplicate(objectSpace.Session);
                    duplicate.Title = duplicateRecord.Key;
                    duplicate.Count = duplicateRecord.Value;
                    collectionSource.List.Add(duplicate);
                }
            }
            ListView view = Application.CreateListView(Application.GetListViewId(typeof(Duplicate)), collectionSource, false);
            view.Editor.AllowEdit = false;
            e.View = view;
            e.DialogController.SaveOnAccept = false;
        }
    }
}

File: VB/UnboundListView.Module/ShowDuplicateBooksController.vb
Imports DevExpress.ExpressApp.DC
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Text

Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base
Imports DevExpress.Xpo

Namespace UnboundListView.Module
	Partial Public Class ShowDuplicateBooksController
		Inherits ViewController
		Public Sub New()
			Dim showDuplicatesAction As New PopupWindowShowAction(Me, "Show Duplicate Books", "View")
			AddHandler showDuplicatesAction.CustomizePopupWindowParams, AddressOf showDuplicatesAction_CustomizePopupWindowParams
			TargetViewType = ViewType.ListView
			TargetObjectType = GetType(Book)
		End Sub

		Private Sub showDuplicatesAction_CustomizePopupWindowParams(ByVal sender As Object, ByVal e As CustomizePopupWindowParamsEventArgs)
			Dim dictionary As New Dictionary(Of String, Integer)()
            For Each book As Book In (CType(Me.View, ListView)).CollectionSource.List
                If (Not String.IsNullOrEmpty(book.Title)) Then
                    If dictionary.ContainsKey(book.Title) Then
                        dictionary(book.Title) += 1
                    Else
                        dictionary.Add(book.Title, 1)
                    End If
                End If
            Next book
			Dim objectSpace As IObjectSpace = Application.CreateObjectSpace()
			Dim collectionSource As New CollectionSource(objectSpace, GetType(Duplicate))
			If (TryCast(collectionSource.Collection, XPBaseCollection)) IsNot Nothing Then
				CType(collectionSource.Collection, XPBaseCollection).LoadingEnabled = False
			End If
			For Each duplicateRecord As KeyValuePair(Of String, Integer) In dictionary
				If duplicateRecord.Value > 1 Then
                    Dim myDuplicate As Duplicate = objectSpace.CreateObject(Of Duplicate)()
                    myDuplicate.Title = duplicateRecord.Key
                    myDuplicate.Count = duplicateRecord.Value
                    collectionSource.List.Add(myDuplicate)
				End If
			Next duplicateRecord
			Dim view As ListView = Application.CreateListView(Application.GetListViewId(GetType(Duplicate)), collectionSource, False)
			view.Editor.AllowEdit = False
			e.View = view
			e.DialogController.SaveOnAccept = False
		End Sub
	End Class
End Namespace
